### PR TITLE
Update: Use new action REPLACE_INNER_BLOCKS for InnerBlocks replacement

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
## Description
Adds a new action REPLACE_INNER_BLOCKS that is used during the inner block creation/update to set its template.
We are updating the insert/replace and move actions to restrict if the new blocks can be added in the specific locations. During the set of an inner blocks template even if restrictions exist e.g: locking, the blocks should be added, this would require flags in the existing actions to not apply restriction in that case.
@gziolo suggested that we use a new specific action, and this PR implements it.

Pending: test cases, after we check if this approach (use a High Order Reducer) is valid I will add them to this PR.

## How has this been tested?
I used the columns block, changed the number of columns, and verified everything still works as before.
I used the Media & Text block and verified the heading block still does not get selected by default.
I tried a sample block with inner blocks https://gist.github.com/jorgefilipecosta/b7194daac3ce26827522694fb4252c1c#file-testallowedblocksmiddleware-js and checked the block still work as before.

### CODE REVIEW

File: packages/editor/src/components/inner-blocks/index.js [hunk 1]
```diff
@@ -144,20 +144,14 @@ InnerBlocks = compose( [
...
 
 		return {
 			replaceInnerBlocks( blocks ) {
-				const clientIds = map( block.innerBlocks, 'clientId' );
-				if ( clientIds.length ) {
-					replaceBlocks( clientIds, blocks );
-				} else {
-					insertBlocks( blocks, undefined, clientId, templateInsertUpdatesSelection );
-				}
+				replaceInnerBlocks( clientId, blocks, block.innerBlocks.length === 0 && templateInsertUpdatesSelection );
```
Comment (gziolo) [inline packages/editor/src/components/inner-blocks/index.js]: `block.innerBlocks.length === 0` - can be computed inside `replaceInnerBlocks`. Is there any reason it has to be placed here?

**Review by gziolo — COMMENTED** [review]
Thanks for updating this PR. I don't quite know how `blocks` reducer works at the moment so I would need to invest some time to understand it better. Seeing the complexity here, it feels like it could use some unit testing for the new action integrated into the reducer.

File: packages/editor/src/components/inner-blocks/index.js (hunk 1 above)
Comment (jorgefilipecosta) [reply packages/editor/src/components/inner-blocks/index.js]: The selection is managed on a reducer that does not have access to the blocks themselves. In order to the check in a reducer, we would need a top-level high order reducer. We don't have any top-level higher-order reducer yet. I'm not sure we want to open that precedent.
But the main reason was to make easier to debug state changes. Currently, if the replaceInnerFlags action updateSelection flag is true the selection is updated if it is false the selection is not updated. If this logic was inside the reducer if the flag was false the selection was not updated, but if the flag was true the selection may or may not be updated.

**Comment by jorgefilipecosta** [comment]
Hi @gziolo that you for the review, I agree that unit tests need to be added, I did not add them yet because I want to validate this approach before investing the time in the test creation.

**Comment by jorgefilipecosta** [comment]
The unit tests were added.

File: packages/block-editor/src/store/test/reducer.js [hunk 2]
```diff
@@ -1562,6 +1904,55 @@ describe( 'state', () => {
...
+
+			expect( state ).toEqual( {
+				start: 'another-block',
+				end: 'another-block',
+				initialPosition: null,
+				isMultiSelecting: false,
+			} );
+		} );
+
+		it( 'should not update the selection on innerblocks replace if updateSelection is false', () => {
```
Comment (gziolo) [inline packages/block-editor/src/store/test/reducer.js]:
```suggestion
		it( 'should not update the selection on inner blocks replace if updateSelection is false', () => {
```

File: packages/block-editor/src/store/test/reducer.js [hunk 3]
```diff
@@ -1562,6 +1904,55 @@ describe( 'state', () => {
 
 			expect( state ).toBe( original );
 		} );
+
+		it( 'should update the selection on innerblocks replace if updateSelection is true', () => {
```
Comment (gziolo) [inline packages/block-editor/src/store/test/reducer.js]:
```suggestion
		it( 'should update the selection on inner blocks replace if updateSelection is true', () => {
```

File: packages/block-editor/src/store/test/reducer.js [hunk 4]
```diff
@@ -162,6 +162,348 @@ describe( 'state', () => {
...
+						'': [ 'clicken' ],
+						clicken: [ newChildBlockId1, newChildBlockId2, newChildBlockId3 ],
+						[ newChildBlockId1 ]: [],
+						[ newChildBlockId2 ]: [],
+						[ newChildBlockId3 ]: [],
+					},
+				} );
+			} );
+
+			it( 'can replace a child block that has other childs', () => {
```
Comment (gziolo) [inline packages/block-editor/src/store/test/reducer.js]:
```suggestion
			it( 'can replace a child block that has other children', () => {
```

File: packages/block-editor/src/store/test/reducer.js [hunk 5]
```diff
@@ -162,6 +162,348 @@ describe( 'state', () => {
 			unregisterBlockType( 'core/test-block' );
 		} );
 
+		describe( 'replace innerblocks', () => {
```
Comment (gziolo) [inline packages/block-editor/src/store/test/reducer.js]:
```suggestion
		describe( 'replace inner blocks', () => {
```

**Review by gziolo — APPROVED** [review]
Thanks for adding unit tests. This gives much more confidence into the newly introduced logic moving forward. Everything looked good in my testing. Great work 👍 

I left some minor proposals for changes in tests names.

**Comment by jorgefilipecosta** [comment]
Thank you for the review and suggestions @gziolo!

**Pull request merged**